modmail/extensions/extension_manager.py

```diff
@@ -0,0 +1,278 @@
+# original source:
+# https://github.com/python-discord/bot/blob/a8869b4d60512b173871c886321b261cbc4acca9/bot/exts/utils/extensions.py  # noqa: E501
+# MIT License 2021 Python Discord
+import functools
+import logging
+import typing as t
+from collections import defaultdict
+from enum import Enum
+
+from discord import AllowedMentions, Colour, Embed
+from discord.ext import commands
+from discord.ext.commands import Context
+
+from modmail.bot import ModmailBot
+from modmail.log import ModmailLogger
+from modmail.utils.cogs import BotModes, ExtMetadata, ModmailCog
+from modmail.utils.extensions import EXTENSIONS, NO_UNLOAD, unqualify
+
+log: ModmailLogger = logging.getLogger(__name__)
+
+
+BASE_PATH_LEN = __name__.count(".")
+
+EXT_METADATA = ExtMetadata(load_if_mode=BotModes.DEVELOP, no_unload=True)
+
+
+class Action(Enum):
+    """Represents an action to perform on an extension."""
+
+    # Need to be partial otherwise they are considered to be function definitions.
+    LOAD = functools.partial(ModmailBot.load_extension)
+    UNLOAD = functools.partial(ModmailBot.unload_extension)
+    RELOAD = functools.partial(ModmailBot.reload_extension)
+
+
+class ExtensionConverter(commands.Converter):
+    """
+    Fully qualify the name of an extension and ensure it exists.
+
+    The * and ** values bypass this when used with the reload command.
+    """
+
+    source_list = EXTENSIONS
+    type = "extension"
+
+    async def convert(self, ctx: Context, argument: str) -> str:
+        """Fully qualify the name of an extension and ensure it exists."""
+        # Special values to reload all extensions
+        if argument == "*" or argument == "**":
+            return argument
+
+        argument = argument.lower()
+
+        if argument in self.source_list:
+            return argument
+
+        qualified_arg = f"modmail.{self.type}s.{argument}"
+        if qualified_arg in self.source_list:
+            return qualified_arg
+
+        matches = []
+        for ext in self.source_list:
+            if argument == unqualify(ext):
+                matches.append(ext)
+
+        if not matches:
+            raise commands.BadArgument(f":x: Could not find the {self.type} `{argument}`.")
+
+        if len(matches) > 1:
+            names = "\n".join(sorted(matches))
+            raise commands.BadArgument(
+                f":x: `{argument}` is an ambiguous {self.type} name. "
+                f"Please use one of the following fully-qualified names.```\n{names}```"
+            )
+
+        return matches[0]
+
+
+class ExtensionManager(ModmailCog, name="Extension Manager"):
+    """Extension management base class."""
```

Comment:
It is not really a base class and is a functional cog itself but a base class for the plugins. So I am not sure if this is a perfect docstring. Maybe you can clear it up :D